### PR TITLE
pdm: 2.10.0 -> 2.10.1; fix #265883

### LINKED ISSUE #265883: pdm 2.10.0 nixpkgs-unstable: permission denied error on `pdm init`
### Describe the bug
When running `pdm init` from `pdm 2.10.0` for creating a new package, the following `permission denied` error occurs. This does not happen with `pdm 2.6.1` currently in `nixpkgs-23.05`.

image

### Steps To Reproduce
1. Use the following flake to get current unstable branch of `pdm 2.10.0` with `python311`
```
{
	inputs = {
		nixpkgs.url = "github:NixOS/nixpkgs/nixos-unstable";
		flake-utils.url = "github:numtide/flake-utils";
	};

	outputs = {nixpkgs, flake-utils, ...} :
		flake-utils.lib.eachSystem [ "x86_64-linux" "aarch64-darwin" ] (system:
			let
				pkgs = import nixpkgs { config.allowUnfree = true; inherit system; };
			in {
			devShells.default = pkgs.mkShell {
				inherit system;
				buildInputs = with pkgs; [ python311 pdm ];
			};
		});
}
```
2. Run `nix develop` to create the shell
3. Run `pdm init` and use defaults - other than specifying that the library is installable.

### Expected behavior
Process should complete without error.

### Notify maintainers

@cpcloud @natsukium 

### Metadata
Please run `nix-shell -p nix-info --run "nix-info -m"` and paste the result.

```console
[user@system:~]$ nix-shell -p nix-info --run "nix-info -m"
 - system: `"x86_64-linux"`
 - host os: `Linux 6.1.57, NixOS, 23.05 (Stoat), 23.05.4335.898cb2064b6e`
 - multi-user?: `yes`
 - sandbox: `yes`
 - version: `nix-env (Nix) 2.13.5`
 - channels(root): `"nixos-23.05"`
 - nixpkgs: `/nix/var/nix/profiles/per-user/root/channels/nixos`
```

### PR DESCRIPTION
**Pull request by natsukium**
## Description of changes

Changelog: https://github.com/pdm-project/pdm/releases/tag/2.10.1

fixes #265883 

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [x] `sandbox = true`
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by giomf** [comment]
Result of `nixpkgs-review pr 266016` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>2 packages built:</summary>
  <ul>
    <li>pdm</li>
    <li>pdm.dist</li>
  </ul>
</details>

**Comment by JonBoyleCoding** [comment]
Fix for #265883 works.

File: pkgs/tools/package-management/pdm/default.nix [hunk 1]
```diff
@@ -98,13 +108,23 @@ buildPythonApplication rec {
...
+        export HOME=$TMPDIR
+        ${lib.getExe pdm} init -n --lib -v >/dev/null
+        install -Dm644 pyproject.toml -t $out
+      '';
+    };
+  };
 
   meta = with lib; {
-    homepage = "https://pdm.fming.dev";
+    homepage = "https://pdm-project.dev";
```
Comment (betaboon) [inline pkgs/tools/package-management/pdm/default.nix:123]:
```suggestion
    homepage = "https://pdm-project.org";
```

Comment (natsukium) [reply pkgs/tools/package-management/pdm/default.nix:123]: Good catch! Thanks.

File: pkgs/tools/package-management/pdm/default.nix [hunk 2]
```diff
@@ -31,15 +33,23 @@ in
...
+    hash = "sha256-0WZTHGWfxJBZM1RlRN0uFs9kjCum2JjIISatakIReoE=";
   };
 
+  # if we use `copy2` to copy files with permission,
+  # we will not be able to write to them because of the read-only nix store
+  # https://github.com/NixOS/nixpkgs/issues/265883
+  postPatch = ''
+    substituteInPlace src/pdm/cli/templates/__init__.py \
+      --replace "shutil.copy2" "shutil.copyfile"
+  '';
```
Comment (mweinelt) [inline pkgs/tools/package-management/pdm/default.nix:51]: Please check with frostming, whether we can get this upstreamed.

Comment (natsukium) [reply pkgs/tools/package-management/pdm/default.nix:51]: reported in https://github.com/pdm-project/pdm/issues/2385

Comment (natsukium) [reply pkgs/tools/package-management/pdm/default.nix:51]: It looks like the PR has already been submitted and merged. https://github.com/pdm-project/pdm/pull/2379
Thanks @adisbladis.

File: pkgs/tools/package-management/pdm/default.nix [hunk 3]
```diff
@@ -31,15 +32,25 @@ in
...
+    hash = "sha256-0WZTHGWfxJBZM1RlRN0uFs9kjCum2JjIISatakIReoE=";
   };
 
+  patches = [
+    # https://github.com/NixOS/nixpkgs/issues/265883
+    # https://github.com/pdm-project/pdm/pull/2379
+    (fetchpatch {
+      name = "fix-template-permission.patch";
+      url = "https://github.com/pdm-project/pdm/commit/f0efdcefe589bc58c28ccf6ce2d23cad9a81dccc.patch";
+      hash = "";
```
Comment (mweinelt) [inline pkgs/tools/package-management/pdm/default.nix:50]:
```suggestion
      hash = "sha256-NnHDSz2N63JzSzh2t9a5f/QQWM6Hyd5Cn5JY2zem6Ac=";
```

Comment (natsukium) [reply pkgs/tools/package-management/pdm/default.nix]: I seem to be a little tired. Thanks.

Comment (mweinelt) [reply pkgs/tools/package-management/pdm/default.nix]: Take care!

File: pkgs/tools/package-management/pdm/default.nix (hunk 2 above)
Comment (adisbladis) [reply pkgs/tools/package-management/pdm/default.nix]: No worries =)

**Pull request merged**